### ISSUE TITLE
Chat participant should get an event if the user interacted with the code in their response

### ISSUE THREAD

**Issue by isidorn** [microsoft/vscode#224476]
Right now we have the `participant.onDidReceiveFeedback` - this is great.
But I would love if extensions could track if the user interacted with the code in their response. I.e. can there be an event in the API that gets triggered whenever a user does a "Insert Code into Editor", "Copy Code", and any other action interaction with the Code in response.

Context: I am writing a "Measuring success" chapter of our Chat docs, and I feel like this is missing to have a compelling story.

![Image](https://github.com/user-attachments/assets/c5f75505-f47e-40cf-b13c-d3314bd4492c)

**Comment by roblourens** [comment]
We do have these events in the "additions" proposal: https://github.com/microsoft/vscode/blob/997ad5efbfe3f33978700c100c0198ad17c01ae7/src/vscode-dts/vscode.proposed.chatParticipantAdditions.d.ts#L307-L310
which we use, and we can consider finalizing some portion of this

**Comment by isidorn** [comment]
Awesome! Yeah, I think it would make the [Measure Success](https://code.visualstudio.com/api/extension-guides/chat#measuring-success) story more complete.

**Comment by roblourens** [comment]
We discussed this in the api sync. The API we have here is very very granular, and it feels like an odd fit for extension API. eg on the copy action:

```
codeBlockIndex: number;
copyKind: ChatCopyKind;
copiedCharacters: number;
totalCharacters: number;
copiedText: string;
```

I'm wondering whether there is something much simpler we could finalize that would still be useful for extensions. What's the minimum we could do? Would an event that says that a copy happened, with no other details, be useful at all? Or, we were thinking about how we can expose some of these details in a more general extension API, rather than something that is just spelling out telemetry events. Something that would have extension authors put the pieces together themselves a bit more, but wouldn't describe these really specific scenarios just for chat. But I'm a little unsure what exactly that would look like, and it would take more time to finalize.

**Comment by isidorn** [comment]
I do not think it needs to be this fine grained. Ideally we define a metric "code interactions" and we recommend extensions authors to measure that for their participants. In addition to `downvotes / total_requests`.
The "code interactions" should be if the user interacted with a code block that your participant replied with. For example, if a participant sent back `10` replies with a code block, and users interacted with `4`, that is a `40%` code interaction metric.

I am fine if we first wait and see how useful this metric is for 1st party participants, before we add it to the API.

**Comment by roblourens** [comment]
Sure, worth thinking about. Off the top of my head I don't know if I would say that interacting with 4/10 codeblocks is less interaction than interacting with 1/1 codeblocks. But saying that it's 4x more interaction also seems wrong so I don't know :slightly_smiling_face:

**Comment by isidorn** [comment]
In our dashboard we will show two metrics for this. Total code blocks and code interactions%. One without the other does not make sense as your example nicely points out.

**Comment by amitjoshi438** [comment]
We're also interested in the code interaction metrics, so thank you for keeping track of them.

**Comment by roblourens** [comment]
@jrieken suggested tracking "interactions" without enumerating what the types of interactions are. Having this + the existing feedback event would be a good way to keep the API lean while giving extension authors something to track, IMO.

**Comment by isidorn** [comment]
Yes, that would work for the scenario I had in mind.

**Comment by jrieken** [comment]
We could have the new kind but also enrich it the feedback-object with the part that has been interacted with. That would be undefined or any of the response parts (which have the same type as `ChatResponseTurn#response`)

```diff
	export enum ChatResultFeedbackKind {
		Unhelpful = 0,
		Helpful = 1,

+.             Interaction = 2
	}


	export interface ChatResultFeedback {
		readonly result: ChatResult;
		readonly kind: ChatResultFeedbackKind;

+              /**
+               * For `Interaction` the response part that has been interacted with
+.              */ 
+             readonly part: undefined | ReadonlyArray<ChatResponseMarkdownPart | ChatResponseFileTreePart | ChatResponseAnchorPart | ChatResponseCommandButtonPart>
	}

```
